services/platform/platform.go

```diff
@@ -112,29 +117,107 @@ func (s *Service) SignUp(ctx context.Context, email, firstName, lastName string)
 	return nil
 }
 
-// SignIn checks Percona Platform user authentication and creates session.
-func (s *Service) SignIn(ctx context.Context, email, password string) error {
-	cc, err := saasdial.Dial(ctx, "", s.host)
+// Connect checks if PMM is connected. If it's not, it connects the a PMM server to the Portal.
+func (s *Service) Connect(ctx context.Context, serverName, email, password string) error {
+	_, err := models.GetPerconaSSODetails(s.db.Querier)
+	if err == nil {
+		return errors.Wrap(err, "PMM server is already connected to Portal")
+	}
+	settings, err := models.GetSettings(s.db)
 	if err != nil {
-		return errors.Wrap(err, "failed establish connection with Percona")
+		return errors.Wrap(err, "failed to fetch PMM server ID and address")
 	}
-	defer cc.Close() //nolint:errcheck
-
-	resp, err := api.NewAuthAPIClient(cc).SignIn(ctx, &api.SignInRequest{Email: email, Password: password})
+	if settings.PMMPublicAddress == "" {
+		return status.Error(codes.FailedPrecondition, "The address of PMM server is not set")
+	}
+	pmmServerURL := fmt.Sprintf("https://%s/graph", settings.PMMPublicAddress)
+	ssoParams, err := s.connect(ctx, &connectPMMParams{
+		serverName:                serverName,
+		email:                     email,
+		password:                  password,
+		pmmServerURL:              pmmServerURL,
+		pmmServerOAuthCallbackURL: fmt.Sprintf("%s/login/generic_oauth", pmmServerURL),
+		pmmServerID:               settings.PMMServerID,
+	})
 	if err != nil {
 		return err
 	}
 
-	err = s.db.InTransaction(func(tx *reform.TX) error {
-		params := models.ChangeSettingsParams{SessionID: resp.SessionId, Email: email}
-		_, err := models.UpdateSettings(tx.Querier, &params)
-		return err
+	err = models.InsertPerconaSSODetails(s.db.Querier, &models.PerconaSSODetails{
+		ClientID:     ssoParams.ClientID,
+		ClientSecret: ssoParams.ClientSecret,
+		IssuerURL:    ssoParams.IssuerURL,
+		Scope:        ssoParams.Scope,
+	})
+	return errors.Wrap(err, "failed to save SSO details")
+}
+
+type connectPMMParams struct {
+	pmmServerURL, pmmServerOAuthCallbackURL, pmmServerID, serverName, email, password string
+}
+
+type connectPMMRequest struct {
```

Comment:
shouldn't https://github.com/percona-platform/platform/blob/becddb1be2e225877a8b3ba372b9a62449837ea9/gen/org/org_api.pb.go#L99 be used instead and response as well?

Reply:
It is a private repository